### PR TITLE
SCVMM Networking Enhancements

### PR DESCRIPTION
**Pull request by agrare**
This adds collection and parsing of VMNetworks and VMSubnets.  VMNetworks are modeled as Lans which have a parent_id.

Based on https://github.com/ManageIQ/manageiq-providers-scvmm/pull/15

https://github.com/ManageIQ/manageiq-schema/pull/69 Added the schema changes and https://github.com/ManageIQ/manageiq/pull/16153 added the base models and save_inventory changes.

### CODE REVIEW

**Comment by bronaghs** [comment]
After this the provisioning dialogs needs to be updated. 
@djberg96 - can the work done as part of the interim automate solution be carried over to this?

**Comment by djberg96** [comment]
@bronaghs Not really, since they're in a different place. Changes to the current provisioning dialogs would have to be made by the UI team, and we'll need to update cloning.rb.

**Comment by bronaghs** [comment]
@djberg96 - The changes to the provisioning dialogs would be done here: https://github.com/ManageIQ/manageiq/blob/master/product/dialogs/miq_dialogs/miq_provision_microsoft_dialogs_template.yaml
Was this file changes as part of the Automate solution?

There might be a small tweak here or there that the UI teams needs to take care of, but for the most part the Providers team can run with this.

**Comment by djberg96** [comment]
@bronaghs The automate solution did not use those dialogs. If those can be used, then that should simplify things hopefully.

If it matters, in the automate solution, the lans, vm networks and subnets are dynamically generated.

File: app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb [hunk 1]
```diff
@@ -193,11 +198,36 @@ def process_virtual_switches(host)
...
+      network_hashes
+    end
+
+    def process_subnets(vm_network)
+      subnets = Array(vm_network['VMSubnet'])
+      subnets.map do |subnet|
+        {
+          :name    => subnet['Name'],
+          :ems_ref => subnet['ID'],
+          :cidr    => subnet['SubnetVLans'],
```
Comment (agrare) [inline app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb]: @djberg96 is `SubnetVLans` a valid CIDR?

I get `["192.168.1.0/24-0", "230", "192.168.31.0/24-231", "241", "231", "192.168.32.0/24-241", "192.168.1.0/24-0", "230", "192.168.31.0/24-231", "241", "231", "192.168.32.0/24-241", "192.168.1.0/24-0"]` which looks like a combination of (I assume) `cidr-vlanid` and just `vlanid`

Comment (djberg96) [reply app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb]: @agrare Yeah, I was confused by that as well when looking at the raw powershell objects. Let me find a few and paste them here.

Comment (djberg96) [reply app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb]: Looks like we're going to have to go a layer deeper:
```
PS C:\> $vmnetwork = Get-SCVMNetwork -Name 'dberger'

PS C:\> $vmnetwork

Name                    : dberger
LogicalNetwork          : rhq
VMSubnet                : {dbergersubnet}
VMNetworkGateways       : {}
VPNConnections          : {}
NATConnections          : {}
RoutingDomainId         : abcdef
IsolationType           : WindowsNetworkVirtualization
UseGRE                  : True
PAIPAddressPoolType     : IPV4
CAIPAddressPoolType     : IPV4
ExternalName            :
NetworkEntityAccessType : VmmManaged
IsAssigned              : True
IsPrivateVlan           : False
HasGatewayConnection    : False
NetworkManager          :
PortACL                 :
GrantedToList           : {}
UserRoleID              : xyz
UserRole                : Administrator
Owner                   : SomeDomain\Administrator
ObjectType              : VMNetwork
Accessibility           : Public
IsViewOnly              : False
Description             : Test VM Network
AddedTime               : 10/9/2017 4:57:36 PM
ModifiedTime            : 8/15/2017 12:21:19 PM
Enabled                 : True
MostRecentTask          :
ServerConnection        : Microsoft.SystemCenter.VirtualMachineManager.Remoting.ServerConnection
ID                      : xxx
MarkedForDeletion       : False
IsFullyCached           : True
MostRecentTaskIfLocal   :

PS C:\> $vmnetwork.vmsubnet

Name                           : dbergersubnet
Description                    :
VMSubnetID                     : 1234567
VMNetwork                      : dberger
SubnetVLans                    : {192.168.1.0/24-0}
AllowsIntraPortCommunication   : True
MaxNumberOfPorts               :
VirtualSwitchExtensionManager  :
LogicalNetworkDefinition       :
ExternalId                     : abc
NetworkEntityAccessType        : VmmManaged
ExternalSyncTime               :
ExternalVMNetworkIsolationType :
ExternalVMSubnetIdentifier     :
IsVMMInternal                  : False
PortACL                        :
ServerConnection               : Microsoft.SystemCenter.VirtualMachineManager.Remoting.ServerConnection
ID                             : xxx
IsViewOnly                     : False
ObjectType                     : VMSubnet
MarkedForDeletion              : False
IsFullyCached                  : True



PS C:\> $vmnetwork.vmsubnet.subnetvlans


Subnet                 : 192.168.1.0/24
VLanID                 : 0
SecondaryVLanID        : 0
IsVLanEnabled          : False
IsSecondaryVLanEnabled : False
SupportsDHCP           : True
ID                     :
IsAssignedToVMSubnet   : True

A number like "240" is just a vlan ID and could be skipped, so I guess we'd have to check the format as well.

**Comment by agrare** [comment]
Okay @djberg96 @bronaghs I updated the inventory json and specs in https://github.com/ManageIQ/manageiq-providers-scvmm/pull/19/commits/80ccb3921cd9f35a6984a045f92c9ef5b8f7bddd and added new spec tests for vm_networks and subnets in the last commit https://github.com/ManageIQ/manageiq-providers-scvmm/pull/19/commits/5348a67aadc52e03544a9189550917942e4ff74e

**Comment by djberg96** [comment]
@agrare I forked the repo, pointed MIQ at it and added an SCVMM provider. There were two possible issues that I noticed. First is that the `parent_id` doesn't appear to be set on the vm networks. Second is that none of the subnets had the `cidr` field set.

Are these things to worry about now? Or is the plan to fix them in a future PR?

**Comment by agrare** [comment]
The parent_id should be saved if you have https://github.com/ManageIQ/manageiq/pull/16165
I thought we were going to do the cidr after you figured out the VLanSubnet/Subnet thing

**Comment by djberg96** [comment]
@agrare I thought I had the latest master, but apparently not. Once I grabbed it, the parent_id was set. We'll work on the cidr thing separately.

**Pull request merged**